File: tests/test_traffic_api.py
import requests
import json
from datetime import datetime

API_KEY = '044cd0c0407245a4bb5d2f6d1d8458bd'
BASE_URL = 'https://openapi.its.go.kr:9443/bypassFCastInfo'

# 현재 날짜와 시간
now = datetime.now()
fcast_date = now.strftime('%Y%m%d')
fcast_hour = now.strftime('%H')

print(f"교통 예측 정보 API 테스트")
print(f"날짜: {fcast_date}, 시간: {fcast_hour}시")
print("=" * 60)

# API 요청 파라미터
params = {
    'apiKey': API_KEY,
    'sectionId': '10',  # 본선 도로 구간 ID
    'fCastDate': fcast_date,
    'fCastHour': fcast_hour,
    'getType': 'json'
}

try:
    print("\nAPI 요청 중...")
    response = requests.get(BASE_URL, params=params)

    print(f"상태 코드: {response.status_code}")

    if response.status_code == 200:
        data = response.json()

        print(f"\n결과 코드: {data['header']['resultCode']}")
        print(f"결과 메시지: {data['header']['resultMsg']}")
        print(f"총 데이터 개수: {data['body']['totalCount']}")

        print("\n=== 도로 구간별 예측 속도 (상위 10개) ===")
        items = data['body']['items'][:10]

        for idx, item in enumerate(items, 1):
            section_type = "메인도로" if item['sectionType'] == 'M' else "우회도로"
            print(f"{idx}. 링크ID: {item['linkId']}")
            print(f"   구간: {section_type}, 길이: {item['length']}m")
            print(f"   예측 속도: {item['speed']}km/h")
            print()

        # 전체 데이터를 JSON 파일로 저장
        with open('traffic_forecast_result.json', 'w', encoding='utf-8') as f:
            json.dump(data, f, ensure_ascii=False, indent=2)

        print("전체 데이터가 'traffic_forecast_result.json' 파일로 저장되었습니다.")

    else:
        print(f"에러 발생: {response.text}")

except Exception as e:
    print(f"예외 발생: {str(e)}")

File: tests/test-traffic-api.js
const https = require('https');

const API_KEY = '044cd0c0407245a4bb5d2f6d1d8458bd';
const BASE_URL = 'https://openapi.its.go.kr:9443/bypassFCastInfo';

// 현재 날짜와 시간 가져오기
const now = new Date();
const fCastDate = now.toISOString().slice(0, 10).replace(/-/g, '');
const fCastHour = String(now.getHours()).padStart(2, '0');

console.log(`테스트 시작 - 날짜: ${fCastDate}, 시간: ${fCastHour}`);

function makeRequest(requestNum) {
  return new Promise((resolve, reject) => {
    const params = new URLSearchParams({
      apiKey: API_KEY,
      sectionId: '10',
      fCastDate: fCastDate,
      fCastHour: fCastHour,
      getType: 'json'
    });

    const url = `${BASE_URL}?${params.toString()}`;

    const startTime = Date.now();

    https.get(url, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        const elapsed = Date.now() - startTime;
        console.log(`\n[요청 #${requestNum}]`);
        console.log(`Status Code: ${res.statusCode}`);
        console.log(`응답 시간: ${elapsed}ms`);

        try {
          const jsonData = JSON.parse(data);
          console.log('응답 데이터:', JSON.stringify(jsonData, null, 2));
          resolve({ success: true, statusCode: res.statusCode, data: jsonData, elapsed });
        } catch (e) {
          console.log('Raw 응답:', data);
          resolve({ success: false, statusCode: res.statusCode, data: data, elapsed });
        }
      });
    }).on('error', (err) => {
      console.error(`[요청 #${requestNum}] 에러:`, err.message);
      reject(err);
    });
  });
}

async function testRateLimit() {
  console.log('\n========================================');
  console.log('API 응답 확인 (단일 요청)');
  console.log('========================================');

  await makeRequest(1);

  console.log('\n========================================');
  console.log('API Rate Limit 테스트 (연속 10회 요청)');
  console.log('========================================');

  const results = [];

  for (let i = 2; i <= 11; i++) {
    try {
      const result = await makeRequest(i);
      results.push(result);

      // 짧은 딜레이
      await new Promise(resolve => setTimeout(resolve, 100));
    } catch (err) {
      results.push({ success: false, error: err.message });
    }
  }

  console.log('\n========================================');
  console.log('테스트 결과 요약');
  console.log('========================================');
  console.log(`총 요청 수: ${results.length + 1}`);
  console.log(`성공: ${results.filter(r => r.success && r.statusCode === 200).length + 1}`);
  console.log(`실패: ${results.filter(r => !r.success || r.statusCode !== 200).length}`);

  const failedRequests = results.filter(r => !r.success || r.statusCode !== 200);
  if (failedRequests.length > 0) {
    console.log('\n실패한 요청:');
    failedRequests.forEach((r, idx) => {
      console.log(`  - 요청 #${idx + 2}: Status ${r.statusCode || 'N/A'}`);
    });
  }
}

testRateLimit();
